### PR TITLE
[SPARK-20574][ML] Allow Bucketizer to handle non-Double numeric column

### PR DESCRIPTION
**Pull request by actuaryzhang**
## What changes were proposed in this pull request?
Bucketizer currently requires input column to be Double, but the logic should work on any numeric data types. Many practical problems have integer/float data types, and it could get very tedious to manually cast them into Double before calling bucketizer. This PR extends bucketizer to handle all numeric types.  

## How was this patch tested?
New test.

### CODE REVIEW

**Comment by actuaryzhang** [comment]
@yinxusen  @srowen @mengxr @jkbradley @VinceShieh @yanboliang 

The example below shows failure of Bucketizer on integer data. 
```
val splits = Array(-3.0, 0.0, 3.0)
val data: Array[Int] = Array(-2, -1, 0, 1, 2)
val expectedBuckets = Array(0.0, 0.0, 1.0, 1.0, 1.0)
val dataFrame = data.zip(expectedBuckets).toSeq.toDF("feature", "expected")
val bucketizer = new Bucketizer()
  .setInputCol("feature")
  .setOutputCol("result")
  .setSplits(splits)
bucketizer.transform(dataFrame)  

java.lang.IllegalArgumentException: requirement failed: Column feature must be of type DoubleType but was actually IntegerType.
```

**Comment by actuaryzhang** [comment]
Weird failure message, the log shows all tests passed...

File: mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala [hunk 1]
```diff
@@ -162,6 +165,28 @@ class BucketizerSuite extends SparkFunSuite with MLlibTestSparkContext with Defa
...
+      .setInputCol("feature")
+      .setOutputCol("result")
+      .setSplits(splits)
+
+    val types = Seq(ShortType, IntegerType, LongType, FloatType)
+    for (mType <- types) {
+      val df = dataFrame.withColumn("feature", col("feature").cast(mType))
+      bucketizer.transform(df).select("result", "expected").collect().foreach {
+        case Row(x: Double, y: Double) =>
+          assert(x === y, "The feature value is not correct after bucketing in type " +
```
Comment (yanboliang) [inline mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala]: ```feature value``` -> ```result```

**Review by yanboliang — COMMENTED** [review]
LGTM except one minor issue. Thanks.

File: mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala [hunk 2]
```diff
@@ -162,6 +165,28 @@ class BucketizerSuite extends SparkFunSuite with MLlibTestSparkContext with Defa
...
+    val data = Array(-2.0, -1.0, 0.0, 1.0, 2.0)
+    val expectedBuckets = Array(0.0, 0.0, 1.0, 1.0, 1.0)
+    val dataFrame: DataFrame = data.zip(expectedBuckets).toSeq.toDF("feature", "expected")
+
+    val bucketizer: Bucketizer = new Bucketizer()
+      .setInputCol("feature")
+      .setOutputCol("result")
+      .setSplits(splits)
+
+    val types = Seq(ShortType, IntegerType, LongType, FloatType)
```
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala]: Other tests for supporting numeric types have included `DecimalType` - often `DecimalType(10, 0)`, as well as `ByteType`. See the various Estimator tests which use `MLTestingUtils.genClassifDFWithNumericLabelCol` and `MLTestingUtils.genRegressionDFWithNumericLabelCol`

**Comment by actuaryzhang** [comment]
Thanks @yanboliang and @MLnick.
I have included the additional numeric types and updated the error msg.

**Comment by yanboliang** [comment]
Merged into master and branch-2.0. Thanks.

**Pull request closed without merging**